File: src/domain/recurrence.ts
import {
  addDays,
  addMonths,
  addYears,
  daysBetween,
  isCalendarDate,
  parseCalendarDate,
} from "./date";
import type { CalendarDate, Occurrence, Operation } from "./types";

export const MAX_OCCURRENCES_PER_OPERATION = 50_000;

function firstCandidateIndex(
  operation: Operation,
  rangeStart: CalendarDate,
): number {
  if (operation.firstDate >= rangeStart) return 0;
  if (operation.recurrence === "daily")
    return Math.max(0, daysBetween(operation.firstDate, rangeStart));
  if (operation.recurrence === "weekly")
    return Math.max(
      0,
      Math.floor(daysBetween(operation.firstDate, rangeStart) / 7),
    );
  const first = parseCalendarDate(operation.firstDate)!;
  const range = parseCalendarDate(rangeStart)!;
  return operation.recurrence === "monthly"
    ? Math.max(
        0,
        (range.year - first.year) * 12 + range.month - first.month - 1,
      )
    : Math.max(0, range.year - first.year - 1);
}

export function occurrenceDates(
  operation: Operation,
  rangeStart: CalendarDate,
  rangeEnd: CalendarDate,
): CalendarDate[] {
  if (
    !operation.enabled ||
    rangeEnd < rangeStart ||
    !isCalendarDate(operation.firstDate)
  )
    return [];
  if (operation.recurrence === "none") {
    return operation.firstDate >= rangeStart && operation.firstDate <= rangeEnd
      ? [operation.firstDate]
      : [];
  }
  const end =
    operation.recurrenceEndDate && operation.recurrenceEndDate < rangeEnd
      ? operation.recurrenceEndDate
      : rangeEnd;
  if (!end || end < operation.firstDate) return [];
  const result: CalendarDate[] = [];
  const initialIndex = firstCandidateIndex(operation, rangeStart);
  for (let offset = 0; offset <= MAX_OCCURRENCES_PER_OPERATION; offset += 1) {
    if (offset === MAX_OCCURRENCES_PER_OPERATION) {
      throw new Error(
        `Серия «${operation.name}» содержит больше ${MAX_OCCURRENCES_PER_OPERATION.toLocaleString("ru-RU")} операций в горизонте.`,
      );
    }
    const index = initialIndex + offset;
    const date =
      operation.recurrence === "daily"
        ? addDays(operation.firstDate, index)
        : operation.recurrence === "weekly"
          ? addDays(operation.firstDate, index * 7)
          : operation.recurrence === "monthly"
            ? addMonths(operation.firstDate, index)
            : addYears(operation.firstDate, index);
    if (!isCalendarDate(date)) break;
    if (date > end) break;
    if (date >= rangeStart) result.push(date);
    if (date === end) break;
  }
  return result;
}

export function expandOccurrences(
  operations: Operation[],
  rangeStart: CalendarDate,
  rangeEnd: CalendarDate,
): Occurrence[] {
  return operations
    .flatMap((operation) =>
      occurrenceDates(operation, rangeStart, rangeEnd).map((date) => ({
        ...operation,
        sourceId: operation.id,
        date,
        occurrenceId: `${operation.id}:${date}`,
        recurring: operation.recurrence !== "none",
      })),
    )
    .sort(
      (left, right) =>
        left.date.localeCompare(right.date) ||
        left.name.localeCompare(right.name, "ru"),
    );
}

File: native/CashFlowCore/Sources/CashFlowCore/Recurrence.swift
import Foundation

public enum RecurrenceEngine {
  public static let maximumOccurrencesPerOperation = 50_000

  public static func defaultEndDate(
    from firstDate: CalendarDate,
    recurrence: Recurrence
  ) throws -> CalendarDate? {
    switch recurrence {
    case .none:
      nil
    case .yearly:
      try firstDate.adding(years: 1)
    case .daily, .weekly, .monthly:
      try firstDate.adding(months: 1)
    }
  }

  public static func nextOccurrenceDate(
    for operation: Operation,
    onOrAfter date: CalendarDate
  ) throws -> CalendarDate? {
    guard operation.enabled else { return nil }
    if operation.recurrence == .none {
      return operation.firstDate >= date ? operation.firstDate : nil
    }

    var index = firstCandidateIndex(for: operation, rangeStart: date)
    var candidate = try occurrenceDate(for: operation, index: index)
    while candidate < date {
      index += 1
      candidate = try occurrenceDate(for: operation, index: index)
    }

    if let endDate = operation.recurrenceEndDate, candidate > endDate { return nil }
    return candidate
  }

  public static func occurrenceDates(
    for operation: Operation,
    from rangeStart: CalendarDate,
    through rangeEnd: CalendarDate
  ) throws -> [CalendarDate] {
    guard operation.enabled, rangeStart <= rangeEnd else { return [] }

    if operation.recurrence == .none {
      return (rangeStart...rangeEnd).contains(operation.firstDate) ? [operation.firstDate] : []
    }

    let seriesEnd = min(operation.recurrenceEndDate ?? rangeEnd, rangeEnd)
    guard seriesEnd >= operation.firstDate else { return [] }

    let initialIndex = firstCandidateIndex(for: operation, rangeStart: rangeStart)
    var result: [CalendarDate] = []
    var index = initialIndex

    while true {
      let date = try occurrenceDate(for: operation, index: index)
      if date > seriesEnd { break }
      if date >= rangeStart {
        guard result.count < maximumOccurrencesPerOperation else {
          throw CashFlowCoreError.tooManyOccurrences(operationID: operation.id)
        }
        result.append(date)
      }
      if date == seriesEnd { break }
      index += 1
    }

    return result
  }

  public static func expand(
    operations: [Operation],
    from rangeStart: CalendarDate,
    through rangeEnd: CalendarDate
  ) throws -> [Occurrence] {
    var result: [Occurrence] = []
    for operation in operations {
      let dates = try occurrenceDates(for: operation, from: rangeStart, through: rangeEnd)
      result.append(contentsOf: dates.map { Occurrence(operation: operation, date: $0) })
    }
    return result.sorted {
      $0.date != $1.date ? $0.date < $1.date : $0.name < $1.name
    }
  }

  private static func firstCandidateIndex(
    for operation: Operation,
    rangeStart: CalendarDate
  ) -> Int {
    guard operation.firstDate < rangeStart else { return 0 }

    switch operation.recurrence {
    case .none:
      return 0
    case .daily:
      return max(0, operation.firstDate.days(until: rangeStart))
    case .weekly:
      return max(0, operation.firstDate.days(until: rangeStart) / 7)
    case .monthly:
      let months =
        (rangeStart.year - operation.firstDate.year) * 12
        + rangeStart.month - operation.firstDate.month
      return max(0, months - 1)
    case .yearly:
      return max(0, rangeStart.year - operation.firstDate.year - 1)
    }
  }

  private static func occurrenceDate(for operation: Operation, index: Int) throws -> CalendarDate {
    switch operation.recurrence {
    case .none:
      operation.firstDate
    case .daily:
      try operation.firstDate.adding(days: index)
    case .weekly:
      try operation.firstDate.adding(days: index * 7)
    case .monthly:
      try operation.firstDate.adding(months: index)
    case .yearly:
      try operation.firstDate.adding(years: index)
    }
  }
}
